src/kv/store.h

```diff
@@ -294,30 +302,40 @@ namespace kv
         {
           LOG_FAIL_FMT("Failed to deserialise snapshot at version {}", v);
           LOG_DEBUG_FMT("No such map in store {}", map_name);
-          success = false;
-          break;
+          return DeserialiseSuccess::FAILED;
         }
 
-        auto map_version = d.deserialise_entry_version();
-        auto map_snapshot = d.deserialise_raw();
+        auto view_search = views.find(map_name);
+        if (view_search != views.end())
+        {
+          LOG_FAIL_FMT("Failed to deserialise snapshot at version {}", v);
+          LOG_DEBUG_FMT("Multiple writes on map {}", map_name);
+          return DeserialiseSuccess::FAILED;
+        }
 
-        search->second->apply_snapshot(map_version, map_snapshot);
-      }
+        auto deserialise_snapshot_view =
+          search->second->deserialise_snapshot(d);
 
-      for (auto& map : maps)
-      {
-        map.second->unlock();
+        // Take ownership of the produced view, store it to be committed
+        // later
+        views[map_name] = {
+          search->second.get(),
+          std::unique_ptr<AbstractTxView>(deserialise_snapshot_view)};
       }
 
-      if (!success)
+      if (!d.end())
       {
+        LOG_FAIL_FMT("Unexpected content in snapshot at version {}", v);
         return DeserialiseSuccess::FAILED;
       }
 
-      if (!d.end())
+      // Each map is committed at a different version, independently of the
+      // snapshot version
+      auto c = apply_views(views, []() { return NoVersion; });
```

Comment:
This seems weird. Why does it work? Aren't we stamping the wrong version into the KV with this?

Reply:
For normal transactions, this lambda returns a version which is applied to the roll of each map affected by the transaction (typically, the current version + 1 for transactions on the source store, or the specific version of the serialised write set when deserialising transactions on target stores). This version is returned by `apply_views()` and then stamped into the KV. 

For snapshots, the version of the (only) element in the roll after the snapshot has been applied differs for each map (some maps in the source store may have been at different versions). So we get that version from the serialised snapshot for each map rather than a common value, hence ignoring the value returned by this lambda entirely. We could write `[]() { return v; }` here and use the return value `c` to set the value of the store but because we don't need to pass `v` to the commit logic, it seems wasteful/could cause confusion? OTOH, if we merge `deserialise_views()` and `deserialise_snapshot()`, we may end up doing that. 

If you think there's a better way to doing this, I'm happy to change this logic.

Reply:
As discussed, this approach is OK. I've updated the comment above to make it clearer where the version applied to each map comes from.